### PR TITLE
Make max size of the memory cache to be configurable.

### PR DESCRIPTION
**Pull request by johnwu96822**
Fix the definition of what 0 expiry time means. In the past we treat 0 expiry time to be no expiry (cache forever) because that's what Dalli does. Now with MemoryStore it treats 0 expiry time as it is. So add code to handle 0 expiry time specifically in cache.rb.

- [ ] @johnny-lai 
- [x] @abdulchaudhrycoupa 
- [x] @supriya

### CODE REVIEW

**Review by abdulchaudhrycoupa — APPROVED** [review]

**Comment by johnwu96822** [comment]
I am going to change the default max memory size to 10MB, since 1 MB (~1500 token verification results) may be too small. But it's just a default value, the application can set it to whatever it wants.

**Comment by johnwu96822** [comment]
```
JohnWu-2018MacBookPro:sand-ruby johnwu$ rspec

Sand::Base
  #cache_key
    returns the cache key
    with either key or scopes being empty
      accepts scopes as empty array or nil
      accepts keys as empty string or nil
  #cache_read
    when key does not exist or data is nil
      returns nil
    when data is expired
      deletes the key from cache and return nil
    when expiry time is 0
      returns the data
    when data has not expired
      returns the data
  #cache_write
    when expires_in_sec > 0
      writes to the cache
    when expires_in_sec <= 0
      writes to the cache with 0 expiry time

Sand::Cache
  should have secondary cache
  #read
    with data only in primary
      returns the value
    with data only in secondary
      returns the value
    without secondary
      returns the value
  #write
    writes to the caches
      returns the value
    without secondary
      returns the value
    expiry_in_sec is 0
      should cache
    expiry_in_sec < 0
      should not cache
  #delete
    removes

Sand::Client
  #request
    with authorized response
      returns the response
    with unknown response code
      returns the response
    with unauthorized response
      with retry
        performs retry and returns 401 response
      with default retry set to 0
        still performs retry once and returns 401 response
      with per-request retry
        performs retry and returns 401 response
  #token
    reading from cache
      uses the token from cache
      with resource not found in cache
        retrieves the token from SAND
    writing to cache
      writes the token to cache
  #oauth_token
    gets token successfully
      returns token and expiry time
    on network error
      without retry
        should raise error and not call sleep
      with retry
        should call sleep and then raise error
      with per-request retry
        should call sleep and then raise error
        with per-request retry set to 0
          should not retry
        with per-request retry set to nil or a negative number
          should default to default_retry_count for number of retries
          should default to default_retry_count for number of retries
  #status_code
    returns nil when given nil

Sand::Service
  #check_request
    with request responds to authorization method
      with valid bearer token
        should return true
      case-insensitive bearer keyword
        should return true
      with invalid bearer token
        should return false because token extracted is empty
      with a blank token
        should return false because token extracted is empty
    with request responds to headers method
      with authorization header and valid bearer token
        should return true
      without authorization header
        should raise an exception
    without any authorization header method
      should raise an exception
  #extract_token
    with valid bearer token
      returns the token
    without bearer token
      returns nil
    without the token
      returns nil
    case-insensitive bearer keyword
      returns the token
  #check_token
    token is empty
      returns false
    Sand responds with 500
      returns allowed => false without caching
    Sand responds with 401
      retries once more
    cache operations
      token and result already cached
        gets the result from cache
      uncached token
        caches token verification result
      expiry time
        token is allowed with exp time
          will compute the expiry time
        token is allowed without exp time
          will use the default expiry time
        token is not allowed
          will use the default expiry time
        without allowed response
          will use the default expiry time
        with invalid allowed response
          will use the default expiry time
  #verify_token
    token is empty
      returns allowed is false
    behaves like a valid sand service
      token allowed
        returns the parsed response body as a hash
      malformed response
        raises JSON parse error
      error response
        raises authentication error
    service does not have a default resource
      and resource was not given
        should raise an ArgumentError
      and resource is passed
        behaves like a valid sand service
          token allowed
            returns the parsed response body as a hash
          malformed response
            raises JSON parse error
          error response
            raises authentication error
    Sand responds with 500
      returns nil
    Sand responds with 401
      raises an error
  #expiry_time
    with future expiry time
      returns expiry time that is not the default
    with expiry time that is passed
      returns a negative number
    with unparsable time
      returns the default expiry time

Finished in 0.11913 seconds (files took 0.45429 seconds to load)
68 examples, 0 failures
```

File: lib/sand/cache.rb [hunk 1]
```diff
@@ -2,19 +2,20 @@
 
 module Sand
   class Cache
+    # Default max size is 10 megabytes
+    MEMORY_MAX_SIZE = 10485760
```
Comment (johnny-lai) [inline lib/sand/cache.rb:6]: Considering you said we can store 10,000 entries in 7M, I feel 10M is too large..

How many tokens are we expected to store for a single hour?

**Comment by stephane-pelchat** [comment]
@johnwu96822 need a justification on the removal for hard dependency on secondary, this is our fallback, it should not be optional

**Comment by johnwu96822** [comment]
@stephane-pelchat it was a mis-communication that made me remove it. I will change it back.

**Comment by johnwu96822** [comment]
After discussing with @johnny-lai, I set the default max cache size to 3 megabytes, which is enough for ~4500 token verification results. This should be sufficient as a default for our Ruby apps and enterprise apps, since we have 2000+ customers. Even if the memory cache is full, the secondary will still kick in.

**Comment by supriya** [comment]
lgtm

File: lib/sand/cache.rb [hunk 2]
```diff
@@ -25,11 +28,19 @@ def read(key)
       value
     end
 
+    # expiry_in_sec = 0 means cache forever.
```
Comment (johnny-lai) [inline lib/sand/cache.rb:31]: Maybe we should not support "forever", or replace "forever" with a long period like 6 hours

File: lib/sand/cache.rb [hunk 3]
```diff
@@ -25,9 +29,11 @@ def read(key)
       value
     end
 
-    def write(key, value, expiry_in_sec = 0)
+    # We modify expiry_in_sec = 0 to cache for 6 hours instead of forever.
+    def write(key, value, expiry_in_sec)
       expiry_in_sec = expiry_in_sec.to_i
-      expiry_in_sec = 0 if expiry_in_sec < 0
+      expiry_in_sec = FOREVER if expiry_in_sec == 0
```
Comment (johnny-lai) [inline lib/sand/cache.rb:35]: Since negative values don't make sense either...
```
expiry_in_sec = FOREVER if expiry_in_sec <= 0
```

**Review by johnny-lai — APPROVED** [review]

**Comment by johnwu96822** [comment]
```
JohnWu-2018MacBookPro:sand-ruby johnwu$ rspec

Sand::Base
  #initialize
    raises ArgumentError when cache is not provided
  #cache_key
    returns the cache key
    with either key or scopes being empty
      accepts scopes as empty array or nil
      accepts keys as empty string or nil
  #cache_read
    when key does not exist or data is nil
      returns nil
    when data is expired
      deletes the key from cache and return nil
    when expiry time is 0
      returns the data
    when data has not expired
      returns the data
  #cache_write
    when expires_in_sec > 0
      writes to the cache
    when expires_in_sec == 0
      writes to the cache with Sand::Cache::FOREVER time
    when expires_in_sec < 0
      does not write to cache

Sand::Cache
  should have secondary cache
  #initialize
    raises ArgumentError when cache is not provided
  #read
    with data only in primary
      returns the value
    with data only in secondary
      returns the value
  #write
    writes to the caches
      returns the value
    expiry_in_sec is 0
      should cache
    expiry_in_sec < 0
      should not cache
  #delete
    removes

Sand::Client
  #request
    with authorized response
      returns the response
    with unknown response code
      returns the response
    with unauthorized response
      with retry
        performs retry and returns 401 response
      with default retry set to 0
        still performs retry once and returns 401 response
      with per-request retry
        performs retry and returns 401 response
  #token
    reading from cache
      uses the token from cache
      with resource not found in cache
        retrieves the token from SAND
    writing to cache
      writes the token to cache
  #oauth_token
    gets token successfully
      returns token and expiry time
    on network error
      without retry
        should raise error and not call sleep
      with retry
        should call sleep and then raise error
      with per-request retry
        should call sleep and then raise error
        with per-request retry set to 0
          should not retry
        with per-request retry set to nil or a negative number
          should default to default_retry_count for number of retries
          should default to default_retry_count for number of retries
  #status_code
    returns nil when given nil

Sand::Service
  #check_request
    with request responds to authorization method
      with valid bearer token
        should return true
      case-insensitive bearer keyword
        should return true
      with invalid bearer token
        should return false because token extracted is empty
      with a blank token
        should return false because token extracted is empty
    with request responds to headers method
      with authorization header and valid bearer token
        should return true
      without authorization header
        should raise an exception
    without any authorization header method
      should raise an exception
  #extract_token
    with valid bearer token
      returns the token
    without bearer token
      returns nil
    without the token
      returns nil
    case-insensitive bearer keyword
      returns the token
  #check_token
    token is empty
      returns false
    Sand responds with 500
      returns allowed => false without caching
    Sand responds with 401
      retries once more
    cache operations
      token and result already cached
        gets the result from cache
      uncached token
        caches token verification result
      expiry time
        token is allowed with exp time
          will compute the expiry time
        token is allowed without exp time
          will use the default expiry time
        token is not allowed
          will use the default expiry time
        without allowed response
          will use the default expiry time
        with invalid allowed response
          will use the default expiry time
  #verify_token
    token is empty
      returns allowed is false
    behaves like a valid sand service
      token allowed
        returns the parsed response body as a hash
      malformed response
        raises JSON parse error
      error response
        raises authentication error
    service does not have a default resource
      and resource was not given
        should raise an ArgumentError
      and resource is passed
        behaves like a valid sand service
          token allowed
            returns the parsed response body as a hash
          malformed response
            raises JSON parse error
          error response
            raises authentication error
    Sand responds with 500
      returns nil
    Sand responds with 401
      raises an error
  #expiry_time
    with future expiry time
      returns expiry time that is not the default
    with expiry time that is passed
      returns a negative number
    with unparsable time
      returns the default expiry time

Finished in 0.12515 seconds (files took 0.53729 seconds to load)
69 examples, 0 failures
```

**Comment by johnwu96822** [comment]
```
JohnWu-2018MacBookPro:sand-ruby johnwu$ rspec

Sand::Base
  #initialize
    raises ArgumentError when cache is not provided
  #cache_key
    returns the cache key
    with either key or scopes being empty
      accepts scopes as empty array or nil
      accepts keys as empty string or nil
  #cache_read
    when key does not exist or data is nil
      returns nil
    when data is expired
      deletes the key from cache and return nil
    when expiry time is 0
      returns the data
    when data has not expired
      returns the data
  #cache_write
    when expires_in_sec > 0
      writes to the cache
    when expires_in_sec == 0
      writes to the cache with Sand::Cache::FOREVER time
    when expires_in_sec < 0
      does not write to cache

Sand::Cache
  should have secondary cache
  #initialize
    raises ArgumentError when cache is not provided
  #read
    with data only in primary
      returns the value
    with data only in secondary
      returns the value
  #write
    writes to the caches
      returns the value
    expiry_in_sec is 0
      should cache
    expiry_in_sec < 0
      should not cache
  #delete
    removes

Sand::Client
  #request
    with authorized response
      returns the response
    with unknown response code
      returns the response
    with unauthorized response
      with retry
        performs retry and returns 401 response
      with default retry set to 0
        still performs retry once and returns 401 response
      with per-request retry
        performs retry and returns 401 response
  #token
    reading from cache
      uses the token from cache
      with resource not found in cache
        retrieves the token from SAND
    writing to cache
      writes the token to cache
  #oauth_token
    gets token successfully
      returns token and expiry time
    on network error
      without retry
        should raise error and not call sleep
      with retry
        should call sleep and then raise error
      with per-request retry
        should call sleep and then raise error
        with per-request retry set to 0
          should not retry
        with per-request retry set to nil or a negative number
          should default to default_retry_count for number of retries
          should default to default_retry_count for number of retries
  #status_code
    returns nil when given nil

Sand::Service
  #check_request
    with request responds to authorization method
      with valid bearer token
        should return true
      case-insensitive bearer keyword
        should return true
      with invalid bearer token
        should return false because token extracted is empty
      with a blank token
        should return false because token extracted is empty
    with request responds to headers method
      with authorization header and valid bearer token
        should return true
      without authorization header
        should raise an exception
    without any authorization header method
      should raise an exception
  #extract_token
    with valid bearer token
      returns the token
    without bearer token
      returns nil
    without the token
      returns nil
    case-insensitive bearer keyword
      returns the token
  #check_token
    token is empty
      returns false
    Sand responds with 500
      returns allowed => false without caching
    Sand responds with 401
      retries once more
    cache operations
      token and result already cached
        gets the result from cache
      uncached token
        caches token verification result
      expiry time
        token is allowed with exp time
          will compute the expiry time
        token is allowed without exp time
          will use the default expiry time
        token is not allowed
          will use the default expiry time
        without allowed response
          will use the default expiry time
        with invalid allowed response
          will use the default expiry time
  #verify_token
    token is empty
      returns allowed is false
    behaves like a valid sand service
      token allowed
        returns the parsed response body as a hash
      malformed response
        raises JSON parse error
      error response
        raises authentication error
    service does not have a default resource
      and resource was not given
        should raise an ArgumentError
      and resource is passed
        behaves like a valid sand service
          token allowed
            returns the parsed response body as a hash
          malformed response
            raises JSON parse error
          error response
            raises authentication error
    Sand responds with 500
      returns nil
    Sand responds with 401
      raises an error
  #expiry_time
    with future expiry time
      returns expiry time that is not the default
    with expiry time that is passed
      returns a negative number
    with unparsable time
      returns the default expiry time

Finished in 0.12515 seconds (files took 0.53729 seconds to load)
69 examples, 0 failures
```

**Comment by johnwu96822** [comment]
[lockit v4.9.3] merged via Lockit CLI on 2021-04-23 15:45:14.663655 -0700 PDT m=+3.438786998

**Pull request merged**